thoughtindustries/helium
# feat: complete signin page

Closes CLM-5968

It seems that there are 3 main components in the helium app: helium graphql endpoint, helium app server, helium app client. User related data (like current user and auth token) is handled from the helium app server side, then the data is hydrated to helium app client side. 

When running helium app in dev environment, the helium app server is an express app which provides a graphql route to proxy all requests to the helium graphql endpoint, a graphiql route for graphql playground, and a catch-all route to handle server-side data fetching to compose page context (some context will be hydrated to helium client side).

When running helium app in production environment (usually the deployed helium app), the helium app server is a cloudflare worker instance. Since user will visit the same TI instance domain, requests are first handled by the core platform to handle server-side data fetching, then the same requests are proxied to the cloudflare worker instance with the server-side data attached in the request header. 

The changes here mainly focus on the helium app in dev environment, since the helium app server behavior is already handled in production environment.

### Changes

- wire up `Login` mutation in sign in page
- changes for helium app in dev environment
  - in proxy `/graphql` route, parse response cookie for `authToken` if set from helium graphql endpoint, then pass-thru the cookie.
  - in catch-all route, parses request cookie header for `authToken` if set, then fetch current user based on the cookie, to initialize page context with the current user and auth token.
  - in proxy `/graphql` route, parse request cookie header for `authToken` if set, then compose the request made to helium graphql endpoint

### Testings

Did some sanity check for the sign-in flow, the general flow is as follows:
1. in incognito window, open the sign in page in helium app
2. sign in with valid user credentials, expect redirection to home page, cookie `authToken` is set in browser, content on home page is specific to the signed-in user.
3. reload home page or other pages (dashboard, catalog), expect that content on the page is specific to the signed-in user.

Verified the above flow for helium app in dev environment and for deployed helium app against TI staging (need query param `preview=<school_preview_pwd>` in TI staging to bypass `school is closed` check).

Also verified the helium app graphiql playground, add `authToken` in request header for user-specific requests. This is changed from the old behavior, where all requests were sent using the user identity of test user email from the ti config.
